File: tests/frontend_sources.py
"""Combined frontend source for tests.

The frontend JS was extracted from the old monolithic ``index.html`` into
``vendor/cvstudio/*.js`` modules. Tests that assert on the frontend source must
therefore look in ``index.html`` (still holds the markup + a small inline init)
PLUS every module. ``frontend_source()`` returns that combined text so an
assertion finds each symbol wherever it was extracted to.
"""

from pathlib import Path

_ROOT = Path(__file__).resolve().parent.parent


def frontend_source() -> str:
    source = (_ROOT / "index.html").read_text(encoding="utf-8-sig")
    module_dir = _ROOT / "vendor" / "cvstudio"
    if module_dir.is_dir():
        for module in sorted(module_dir.glob("*.js")):
            source += "\n" + module.read_text(encoding="utf-8")
    return source

File: tests/frontend_sources.js
'use strict';

// The frontend JS was extracted from the old monolithic index.html into
// vendor/cvstudio/*.js modules. Tests that search the frontend source for a
// function/definition/string must look in index.html (still holds the markup +
// a small inline init) PLUS the modules it loads.
//
// The modules are read in the SAME ORDER as index.html's <script src> tags, and
// ONLY the files actually referenced there — not an alphabetical directory
// listing. Following the real load order means a test can only find a symbol if
// the browser actually loads the file that defines it; an orphaned or unused
// module file can never produce a false pass.
const fs = require('fs');
const path = require('path');

const ROOT = path.resolve(__dirname, '..');
const MODULE_TAG = /<script[^>]+src=["']\/vendor\/cvstudio\/([A-Za-z0-9._-]+\.js)(?:\?[^"']*)?["']/g;

function frontendSource() {
  const indexHtml = fs.readFileSync(path.join(ROOT, 'index.html'), 'utf8');
  let source = indexHtml;
  const seen = new Set();
  let match;
  while ((match = MODULE_TAG.exec(indexHtml))) {
    const name = match[1];
    if (seen.has(name)) continue;
    seen.add(name);
    const file = path.join(ROOT, 'vendor', 'cvstudio', name);
    if (fs.existsSync(file)) source += '\n' + fs.readFileSync(file, 'utf8');
  }
  return source;
}

module.exports = { frontendSource, ROOT };
